src/dissect/dissect.c

```diff
@@ -370,6 +432,170 @@ static int run(int argc, char *argv[]) {
+
+                                r = copy_bytes(STDIN_FILENO, target_fd, (uint64_t) -1, COPY_REFLINK);
+                                if (r < 0)
+                                        return log_error_errno(r, "Failed to copy bytes from stdin to '%s' in image '%s': %m", arg_target, arg_image);
+
+                                /* When we copy from stdin we don't copy any attributes (i.e. no access mode, no ownership, no xattr, no times) */
+                                break;
+                        }
+
+                        source_fd = open(arg_source, O_RDONLY|O_CLOEXEC|O_NOCTTY);
+                        if (source_fd < 0)
+                                return log_error_errno(source_fd, "Failed to open source path '%s': %m", arg_source);
+
+                        r = fd_verify_regular(source_fd);
+                        if (r < 0) {
+                                if (r != -EISDIR)
+                                        return log_error_errno(r, "Source '%s' is neither regular file nor directory: %m", arg_source);
+
+                                /* We are looking at a directory. */
+
+                                target_fd = openat(dfd, basename(arg_target), O_RDONLY|O_DIRECTORY|O_CLOEXEC);
+                                if (target_fd < 0) {
+                                        if (errno != ENOENT)
+                                                return log_error_errno(errno, "Failed to open destination '%s': %m", arg_target);
+
+                                        r = copy_tree_at(source_fd, ".", dfd, basename(arg_target), UID_INVALID, GID_INVALID, COPY_REFLINK|COPY_REPLACE|COPY_SIGINT);
+                                } else
+                                        r = copy_tree_at(source_fd, ".", target_fd, ".", UID_INVALID, GID_INVALID, COPY_REFLINK|COPY_REPLACE|COPY_SIGINT);
+                                if (r < 0)
+                                        return log_error_errno(r, "Failed to copy '%s' to '%s' in image '%s': %m", arg_source, arg_target, arg_image);
+
+                                break;
+                        }
+
+                        /* We area looking at a regular file */
+                        target_fd = openat(dfd, basename(arg_target), O_WRONLY|O_CREAT|O_CLOEXEC|O_NOCTTY|O_EXCL, 0600);
+                        if (target_fd < 0)
+                                return log_error_errno(errno, "Failed to open target file '%s': %m", arg_target);
+
+                        r = copy_bytes(source_fd, target_fd, (uint64_t) -1, COPY_REFLINK);
+                        if (r < 0)
+                                return log_error_errno(r, "Failed to copy bytes from '%s' to '%s' in image '%s': %m", arg_source, arg_target, arg_image);
+
+                        (void) copy_xattr(source_fd, target_fd);
+                        (void) copy_access(source_fd, target_fd);
+                        (void) copy_times(source_fd, target_fd, 0);
```

Comment:
Hmm, so I though about this, and I am very sure we should not copy owners, because that clearly makes only sense within an image's context, hence I didn't do that. But whether a file is RO or not and auxiliary info in xattrs I figured is something that shouldn't be lost on the way... I mean, copying a file like /etc/shadow from an image to the host should probably not open it up to everyone, but if that's desired require a separate step...

Noone knows what xattrs contain, but I'd assume that the user xattrs (which is what this is about) are generally user meta information, and there's little reason to belive it makes no sense outside of the image?

Reply:
copying xattrs also is quite handy when a file is already selinux-labeled, and you want to avoid having to re-label it after the fact